fix(ivy): unable to bind to implicit receiver in embedded views

devversion opened the pull request:
To provide some context: The implicit receiver is part of the
parsed Angular template AST. Any property reads in bindings,
interpolations etc. read from a given object (usually the component
instance). In that case there is an _implicit_ receiver which can also
be specified explicitly by just using `this`.

e.g.

```html
<ng-template>{{this.myProperty}}</ng-template>
```

This works as expected in Ivy and View Engine, but breaks in case the
implicit receiver is not used for property reads. For example:

```html
<my-dir [myFn]="greetFn.bind(this)"></my-dir>
```

In that case the `this` will not be properly translated into the generated
template function code because the Ivy compiler currently always treats
the `ctx` variable as the implicit receiver. This is **not correct** and breaks
compatibility with View Engine. Rather we need to ensure that we retrieve
the root context for the standalone implicit receiver similar to how it works
for property reads (as seen in the example above with `this.myProperty`)

Note that this requires some small changes to the `expression_converter`
because we only want to generate the `eenextContent()` instruction if the
implicit receiver is _actually_ used/needed. View Engine determines if that is the case
by recursively walking through the converted output AST and checking for usages of
the `o.variable('_co')` variable ([see here][ve_check]). This would work too for Ivy, but involves
most likely more code duplication since templates are isolated in different functions and it
another pass through the output AST for every template expression.

[ve_check]: https://github.com/angular/angular/blob/0d6c9d36a174f7dc6eb1029e459beecc2dfb0026/packages/compiler/src/view_compiler/view_compiler.ts#L206-L208

Resolves FW-1366.

AndrewKushnir reviewed (approved):
Looks great Paul, thanks for the fix 👍 

I think it'd be great if @alxhub take a look at at as well.

devversion committed:
fix(ivy): unable to bind to implicit receiver in embedded views

To provide some context: The implicit receiver is part of the
parsed Angular template AST. Any property reads in bindings,
interpolations etc. read from a given object (usually the component
instance). In that case there is an _implicit_ receiver which can also
be specified explicitly by just using `this`.

e.g.

```html
<ng-template>{{this.myProperty}}</ng-template>
```

This works as expected in Ivy and View Engine, but breaks in case the
implicit receiver is not used for property reads. For example:

```html
<my-dir [myFn]="greetFn.bind(this)"></my-dir>
```

In that case the `this` will not be properly translated into the generated
template function code because the Ivy compiler currently always treats
the `ctx` variable as the implicit receiver. This is **not correct** and breaks
compatibility with View Engine. Rather we need to ensure that we retrieve
the root context for the standalone implicit receiver similar to how it works
for property reads (as seen in the example above with `this.myProperty`)

Note that this requires some small changes to the `expression_converter`
because we only want to generate the `eenextContent()` instruction if the
implicit receiver is _actually_ used/needed. View Engine determines if that is the case by recursively walking through the converted output AST and
checking for usages of the `o.variable('_co')` variable ([see here][ve_check]). This would work too for Ivy, but involves most likely more code duplication
since templates are isolated in different functions and it another pass
through the output AST for every template expression.

[ve_check]: https://github.com/angular/angular/blob/0d6c9d36a174f7dc6eb1029e459beecc2dfb0026/packages/compiler/src/view_compiler/view_compiler.ts#L206-L208

Resolves FW-1366.

JoostK commented:
Looks like this fixes #30507 as well :-)

alxhub reviewed (approved):
Nice work, sir!

AndrewKushnir commented:
- [VE Presubmit](http://test/OCL:252508372:BASE:252508395:1560208897904:3f1534fd)
- [Ivy Presubmit](https://test.corp.google.com/ui#id=OCL:252508681:BASE:252508990:1560209095017:dafa3791)

AndrewKushnir commented:
@devversion  merging this PR to patch branch failed due to some conflicts. I've changed the label to "master-only" and merged it to master. Could you please create a separate PR and target patch branch? Thank you.
